NDTensors/src/dense/dense.jl

```diff
@@ -3,70 +3,101 @@
 #
 using LinearAlgebra: BlasFloat
 
-struct Dense{ElT,VecT<:AbstractArray} <: TensorStorage{ElT}
-  data::VecT
-  function Dense{ElT,VecT}(data::AbstractArray) where {ElT,VecT<:AbstractArray{ElT}}
-    return new{ElT,VecT}(data)
+struct Dense{ElT,DataT<:AbstractArray} <: TensorStorage{ElT}
+  data::DataT
+  function Dense{ElT,DataT}(data::AbstractArray) where {ElT,DataT<:AbstractArray{ElT}}
+    return new{ElT,DataT}(data)
   end
 
-  # Special case of Vector
-  function Dense{ElT,VecT}(data::Vector) where {ElT,VecT<:AbstractArray{ElT}}
-    return new{ElT,VecT}(data)
-  end
+  # # Special case of Vector
+  # function Dense{ElT,DataT}(data::Vector) where {ElT,DataT<:AbstractArray{ElT}}
+  #   return new{ElT,DataT}(data)
+  # end
 
   # Special case of Array which gets turned into a Vector
-  function Dense{ElT,VecT}(data::Array) where {ElT,VecT<:AbstractArray{ElT}}
-    return new{ElT,VecT}(vec(data))
-  end
+  # function Dense{ElT,DataT}(data::Array) where {ElT,DataT<:AbstractArray{ElT}}
+  #   return new{ElT,DataT}(vec(data))
+  # end
+end
+
+#Start with high information constructors and move to low information constructors
+function Dense{ElT,DataT}() where {ElT,DataT<:AbstractArray{ElT}}
+  return Dense{ElT,DataT}(DataT())
+end
+
+# Construct from a set of indices
+# This will fail if zero(ElT) is not defined for the ElT
+function Dense{ElT,DataT}(inds::Tuple) where {ElT,DataT<:AbstractArray{ElT}}
+  return Dense{ElT,DataT}(zeros(DataT, dim(inds)))
+end
+
+function Dense{ElT,DataT}(
+  ::UndefInitializer, inds::Tuple
+) where {ElT,DataT<:AbstractArray{ElT}}
+  return Dense{ElT,DataT}(DataT(undef, dim(inds)))
```

Comment:
I guess there will be a problem in functions like this where if `DataT` is not `AbstractVector` (i.e. a more general `AbstractArray`) then this construct would fail since you only provide a single dimension. For example:
```julia
julia> Array{Int,3}(undef, 4)
ERROR: MethodError: no method matching Array{Int64, 3}(::UndefInitializer, ::Int64)
Closest candidates are:
  Array{T, 3}(::UndefInitializer, ::Int64, ::Int64, ::Int64) where T at boot.jl:463
  Array{T, N}(::UndefInitializer, ::Int64...) where {T, N} at boot.jl:465
  Array{T, N}(::UndefInitializer, ::Integer...) where {T, N} at baseext.jl:25
  ...
Stacktrace:
 [1] top-level scope
   @ REPL[4]:1
```
A fix could be to use `dims(inds)` instead, which would require that `length(inds) == ndims(DataT)`.

Then we could have another version that makes `length(inds) == ndims(DataT)` if they are not already equal. But this is a tricky case now that we're thinking of making `DataT` more general.

Reply:
I would imagine it would be a good idea to throw an error in general if length(inds) != ndims(DataT) because it is possible for us to cut inds up to make something the right size but we can't in general know what the right inds will be

Reply:
Yeah I think that's right. The `DataT` should get fixed to the correct `ndims` in higher up functions based on `inds` (or something else) before it gets to this function, say from constructors that don't specify the data type like `Dense{Float64}(undef, (3, 3, 3))`. That could generically get handled by defining `default_datatype(inds::Tuple)` for example.